### DISCUSSION TITLE
Weighted mean quaternion/orientation

### DISCUSSION THREAD

**Discussion by DorianDepriester** [pyxem/orix#435] [Ideas]
Hello,
I was looking for a way to compute the _weighted_ mean orientation of crystals. Starting from the discussion in issue #434, and mixing it with [the current implementation](https://github.com/pyxem/orix/blob/c7af446c625f908f9f077e0f36386e5f54322523/orix/quaternion/quaternion.py#L679-L696) of the mean method(), together with the the full Eq. (13) in the given [reference](https://arc.aiaa.org/doi/pdf/10.2514/1.28949), I ended with this:
```python
def weighted_mean(o, weights):
  o2 = o.map_into_symmetry_reduced_zone()
  q = o2.data.T
  qq = (weights*q).dot(q.T)
  w, v = np.linalg.eig(qq)
  w_max = np.argmax(w)
  qmean = v[:, w_max]
  return Orientation(Quaternion(qmean), o.symmetry)
```
I am probably wrong (that's why I haven't made a pull request). Anyway, I think this feature could be useful (at least for me :wink:).

**Comment by hakonanes** [comment]
Thanks for this Dorian, this would be very useful indeed!

Started testing this a bit with the following procedure (in an IPython console), and it looks promising to me. To probe the weighting, I generated a grid of six orientations, +/- 30 degrees about each sample direction x, y and z, and then plotted mean orientations with different weights, like so: 

```python
%matplotlib qt5
import numpy as np
from orix.quaternion import Orientation, Quaternion, symmetry


def weighted_mean(o, weights):
    o2 = o.map_into_symmetry_reduced_zone()
    q = o2.data.T
    qq = (weights*q).dot(q.T)
    w, v = np.linalg.eig(qq)
    w_max = np.argmax(w)
    qmean = v[:, w_max]
    return Orientation(Quaternion(qmean), o.symmetry)

ori = Orientation.from_axes_angles(
    [[1, 0, 0], [1, 0, 0], [0, 1, 0], [0, 1, 0], [0, 0, 1], [0, 0, 1]],
    [30, -30, 30, -30, 30, -30],
    symmetry=symmetry.Oh,
    degrees=True
)

fig = ori.scatter(return_figure=True, c=np.arange(ori.size))
ax = fig.axes[0]
ax.set(xlabel="x [rad]", ylabel="y [rad]", zlabel="z [rad]")
ax.axis(True)
ax.grid(False)

# Same as mean: red
ori_w1 = weighted_mean(ori, [1, 1, 1, 1, 1, 1])
ax.scatter(ori_w1, c="r")

# Weight 30deg [100] 10 times more than the others: blue
ori_w2 = weighted_mean(ori, [10, 1, 1, 1, 1, 1])
ax.scatter(ori_w2, c="b")

# Weight 30deg about [100] and [001] 10 times more than the others: cyan
ori_w3 = weighted_mean(ori, [10, 1, 1, 1, 10, 1])
ax.scatter(ori_w3, c="c")

# Save different views of axis-angle space (using fig.savefig(), not shown)
ax.view_init(90, -90)  # XY plane
ax.view_init(0, -90)  # -XZ plane
```

![ori_weight_30_-60](https://user-images.githubusercontent.com/12139781/225016528-f18458dd-89fc-4208-8b0e-beb4c2a59404.png)|![ori_weight_90_-90](https://user-images.githubusercontent.com/12139781/225016620-7c6f4d3e-ae2e-43d8-95ea-a9bd1ea177c7.png)
:---:|:---:
|![ori_weight_0_-90](https://user-images.githubusercontent.com/12139781/225019855-f22c9cdb-58a9-47e6-affd-382bd0c4ab39.png)

We must supplement this visual test with some quantitative ones... Once those are found, we should add this feature (with a PR from you if you're willing).

**Reply by hakonanes** [reply]
Actually, I guess a valid test is just to compare the mean when having two of the same orientations in a set, or just having one but weighting it twice. If such a test passes, I'm happy with including this feature.

**Reply by DorianDepriester** [reply]
I agree. Your previous test actually seems robust to me!

**Comment by DorianDepriester** [comment]
Yet, as another step toward complexity, let's assume that one wants to use a matrix of weights. The following function seems to work without any loop:
```python
def weighted_mean_matrix(o, mat):
    o2 = o.map_into_symmetry_reduced_zone()
    q = o2.data
    qq = np.einsum('pi,ij,ik->pjk', mat, q, q)
    w, v = np.linalg.eig(qq)
    w_max = np.argmax(w, axis=1)
    q_mean= v[np.arange(mat.shape[0]), :, w_max]
    return Orientation(Quaternion(q_mean), o.symmetry)
```

Again, this should be tested (as made above)

**Reply by hakonanes** [reply]
By matrix, do you mean that your orientations have `ndim` > 1? The einsum you use works for 2D, but how about 3D? Another solution is to reshape `q` to shape (n, 4) and ravel `weights`, and then reshape `q_mean` again before returning, right?

**Reply by DorianDepriester** [reply]
Consider _n_ orientations (`o.shape=(n,)`). If one wants to compute _m_ weighted mean orientations, the matrix will be of size (_m_,_n_). It basically does the same as the an iteration of the function defined above.

**Reply by hakonanes** [reply]
Ah, I see, sorry for my misunderstanding. Your generalization seems to cover all cases, then! In the implementation we should flatten the orientations before applying the weights. After that I think this should be robust.
